src/utils.jl

```diff
@@ -107,29 +107,88 @@ isdef(ex) = ismatch(or_(:(function _(__) _ end),
                         :(f_(__) = _)),
                     ex)
 
-function longdef(ex)
-  prewalk(ex) do ex
-    @match ex begin
-      (f_(args__) = body_) => @q function $f($(args...)) $body end
-      ((args__,) -> body_) => @q function ($(args...),) $body end
-      (arg_ -> body_) => @q function ($arg,) $body end
-      _ => ex
-    end
+function longdef1(ex)
+  @match ex begin
+    (f_(args__) = body_) => @q function $f($(args...)) $body end
+    (f_(args__)::rtype_ = body_) => @q function $f($(args...))::$rtype $body end
+    ((args__,) -> body_) => @q function ($(args...),) $body end
+    (arg_ -> body_) => @q function ($arg,) $body end
+    _ => ex
+  end
+end
+longdef(ex) = prewalk(longdef1, ex)
+
+function shortdef1(ex)
+  @match ex begin
+    function f_(args__) body_ end => @q $f($(args...)) = $body
+    function f_(args__)::rtype_ body_ end => @q $f($(args...))::$rtype = $body
+    function (args__,) body_ end => @q ($(args...),) -> $body
+    ((args__,) -> body_) => ex
+    (arg_ -> body_) => @q ($arg,) -> $body
+    _ => ex
   end
 end
+shortdef(ex) = prewalk(shortdef1, ex)
+
+""" `split_kwargs(x)` splits an argument list into positional and keyword args.
+Returns `(args::Vector, kwargs::Vector)` """
+function split_kwargs(args)
+    positional_args = []
+    kwargs = []
+    for x in args
+        if isa(x, Expr) && x.head == :parameters
+            @assert kwargs == []
+            kwargs = x.args
+        else
+            push!(positional_args, x)
+        end
+    end
+    return (positional_args, kwargs)
+end
+
+"""    parsedef(fdef)
+
+Match a function definition such as
+
+```julia
+function fname(args; kwargs)::return_type
+   body_block
+end
+```
+
+and returns `(fname::Symbol, args::Vector{Any}, kwargs::Vector{Any}, body_block::Expr, return_type)`. `return_type` is `:Any` if not specified. """
+function parsedef(fdef)
+    @match longdef1(fdef) begin
+        (function fname_(args__) body_ end =>
+         (fname, split_kwargs(args)..., body, :Any))
+        (function fname_(args__)::rtype_ body_ end =>
+         (fname, split_kwargs(args)..., body, rtype))
+        any_ => error("Not a function definition: $fdef")
+    end
+end
+
 
-function shortdef(ex)
-  prewalk(ex) do ex
-    @match ex begin
-      function f_(args__) body_ end => @q $f($(args...)) = $body
-      function (args__,) body_ end => @q ($(args...),) -> $body
-      ((args__,) -> body_) => ex
-      (arg_ -> body_) => @q ($arg,) -> $body
-      _ => ex
+""" `parsearg(arg)` matches function arguments (whether from a definition or a function
+call) such as `x::Int=2` and returns `(arg_name, arg_type, default)`. For example:
+ - `parsearg(parsedef(:(f(x::Int=2)=3))[2][1]) -> (:x, :Int, Nullable(2))`.
+ - `parsearg(:x) -> (:x, :Any, Nullable())`
+"""
+function parsearg(arg_expr)
+    if isa(arg_expr, Expr) && arg_expr.head == :kw
+        default = Nullable(arg_expr.args[2])
+        arg_expr = arg_expr.args[1]
+    else
+        default = Nullable()
```

Comment:
In `f(x=nothing)`, `:nothing` would be a symbol, so that wouldn't be an issue. You could technically write `f(x=$nothing)` and break it though, so maybe nullable is the right route.